typing/ctype.ml

```diff
@@ -5200,9 +5200,13 @@ let rec nondep_type_rec ?(expand_private=false) env ids ty =
             | _ -> Tvariant row
           end
       | desc -> copy_type_desc (nondep_type_rec env ids) desc
-    in
-    Transient_expr.set_stub_desc ty' desc;
-    ty'
+    with
+    | desc ->
+      Transient_expr.set_stub_desc ty' desc;
+      ty'
+    | exception e ->
+      TypeHash.remove nondep_hash ty;
+      raise e
```

Comment:
This looks like a use-case for `Misc.try_finally ~exceptionally:(...) @@ fun () ->`

Reply:
Honestly, that seems much less clear than match-with-exception!

Reply:
I disagree, the benefit of my proposal is that the cleanup code is close to the setup code, instead of being at the very end of the code, making the connection between the two harder to see.
But: it's your code, so let's go with your pet preferences instead of mine.

Reply:
The invariant we want is that every stub created is either resolved or removed. match-with-exception puts stub resolution right beside stub removal, and it's easy to see that exactly one of them happens. Your proposal puts stub removal right beside stub creation (which is nice!), but makes it harder to see that either resolution or removal occurs, because resolution remains far away. (IOW: it's nice to have setup and cleanup together, but I don't like having setup and half of cleanup together with the other half far away).

Reply:
*/me goes writing a RFC about `with <handler> match <expr>`*